### PR TITLE
Local Area graphs

### PR DESCRIPTION
**Pull request by JPXKQX**
This PR implements the functionality to build graphs only in an area of interest (AOI).

This includes: 
- A new node builder, **CutOutZarrDatasetNodes**, which allows to create nodes from 2 zarr datasets using cutout operation from [anemoi-datasets](https://anemoi-datasets.readthedocs.io/en/latest/using/combining.html#cutout).
- A new class, **KNNAreaMaskBuilder**, to specify an AOI based on a set of nodes, for which we can specify node attributes to mask the nodes of interest.
- A new set of node builders, **LimitedArea...Nodes**, which allows the creation of the nodes only within an Area Of Interest (AOI).
- Expanding the **MultiScaleEdges** functionalities to support multi-scale connections in limited area graphs.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated the documentation and docstrings to reflect the changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured that the code is still pip-installable after the changes and runs
- [x] I have ran this on single GPU
- [ ] I have ran this on multi-GPU or multi-node

## Issue ticket number and link

Issues #7 

This work would not have been possible without @theissenhelen  and @JesperDramsch and the rest of the AIFS team!

----
📚 Documentation preview 📚: https://anemoi-graphs--30.org.readthedocs.build/en/30/

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=30) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=30) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: JPXKQX<br/>:x: theissenhelen<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-graphs?pullRequest=30) it.</sub>

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=30) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=30) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: theissenhelen<br/>:x: JPXKQX<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-graphs?pullRequest=30) it.</sub>

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=30) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-graphs?pullRequest=30) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: theissenhelen<br/>:x: JPXKQX<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-graphs?pullRequest=30) it.</sub>

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,11 +10,17 @@ Keep it human-readable, your future self will thank you!
 
 ## [Unreleased](https://github.com/ecmwf/anemoi-graphs/compare/0.3.0...HEAD)
 
+### Added
+- New node builder class, CutOutZarrDatasetNodes, to create nodes from 2 datasets.
```
Comment (mchantry) [inline CHANGELOG.md:14]: Can we add a tag to the PR please for these lines. i.e. #30

File: src/anemoi/graphs/generate/hexagonal.py [hunk 2]
```diff
@@ -3,79 +3,78 @@
...
 
 
 def get_nodes_at_resolution(
     resolution: int,
-    area: dict | None = None,
-) -> set[str]:
+) -> list[str]:
     """Get nodes at a specified refinement level over the entire globe.
 
     If area is not None, it will return the nodes within the specified area
```
Comment (mchantry) [inline src/anemoi/graphs/generate/hexagonal.py:80]: Does this need removing if area is no longer here?

File: src/anemoi/graphs/generate/hexagonal.py [hunk 3]
```diff
@@ -121,6 +114,8 @@ def add_edges_to_nx_graph(
     depth_children : int
         The number of resolution levels to consider for the connections of children. Defaults to 1, which includes
```
Comment (mchantry) [inline src/anemoi/graphs/generate/hexagonal.py:115]: Default has changed?

File: src/anemoi/graphs/generate/icosahedral.py [hunk 4]
```diff
@@ -5,50 +5,54 @@
...
 from sklearn.neighbors import BallTree
 
+from anemoi.graphs.generate.masks import KNNAreaMaskBuilder
 from anemoi.graphs.generate.transforms import cartesian_to_latlon_rad
+from anemoi.graphs.generate.utils import get_coordinates_ordering
 
 
 def create_icosahedral_nodes(
-    resolutions: list[int],
+    resolution: int, aoi_mask_builder: KNNAreaMaskBuilder | None = None
```
Comment (mchantry) [inline src/anemoi/graphs/generate/icosahedral.py:16]: why is resolutions changing from list to int?

File: src/anemoi/graphs/generate/icosahedral.py [hunk 5]
```diff
@@ -59,19 +63,11 @@ def create_icosahedral_nx_graph_from_coords(coords_rad: np.ndarray, node_orderin
...
-    index_longitude = np.argsort(coords_rad[index_latitude][:, 0])[::-1]
-    node_ordering = np.arange(coords_rad.shape[0])[index_latitude][index_longitude]
-    return node_ordering
-
-
 def add_edges_to_nx_graph(
     graph: nx.DiGraph,
     resolutions: list[int],
     x_hops: int = 1,
+    aoi_mask_builder: KNNAreaMaskBuilder | None = None,
```
Comment (mchantry) [inline src/anemoi/graphs/generate/icosahedral.py:70]: What is `aoi` please?

**Review by mchantry — COMMENTED** [review]
Looking good. Haven't run the tests yet.

File: src/anemoi/graphs/generate/hexagonal.py (hunk 3 above)
Comment (JPXKQX) [reply src/anemoi/graphs/generate/hexagonal.py:115]: With x_hops=1, children of the same node will be connected. So, depth_children=0 & 1 will produce the same graph.

File: src/anemoi/graphs/generate/icosahedral.py (hunk 4 above)
Comment (JPXKQX) [reply src/anemoi/graphs/generate/icosahedral.py:16]: In both hex & tri refinements, all nodes from one level are contained in the next refinement level. Before, we were getting the maximum of the list.

PS: The list of resolutions is still needed to define the scales at which to build the connections.

File: src/anemoi/graphs/generate/icosahedral.py (hunk 5 above)
Comment (JPXKQX) [reply src/anemoi/graphs/generate/icosahedral.py:70]: it refers to "area of interest", but I think it is better to rename it to area_mask_builder.

File: src/anemoi/graphs/edges/builder.py [hunk 6]
```diff
@@ -9,13 +9,14 @@
...
 from hydra.utils import instantiate
+from scipy.sparse import coo_matrix
 from sklearn.neighbors import NearestNeighbors
 from torch_geometric.data import HeteroData
 from torch_geometric.data.storage import NodeStorage
 
 from anemoi.graphs import EARTH_RADIUS
-from anemoi.graphs.generate import hexagonal
-from anemoi.graphs.generate import icosahedral
+from anemoi.graphs.generate import hex_icosahedron
```
Comment (mchantry) [inline src/anemoi/graphs/edges/builder.py:18]: nice!

**Review by mchantry — APPROVED** [review]
Great work all involved.

**Review by mchantry — APPROVED** [review]
Looks good to me.
